### PR TITLE
[ONNX] Use decorators for symbolic function registration

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85350
* #85349
* #85312
* #85165
* #84776
* #84728
* #85290
* __->__ #84448

This is the 4th PR in the series of #83787. It enables the use of `@onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `@onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `@onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
    - Remove the outdated unit test `test_symbolic_opset9.py`
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  image

### CODE REVIEW

File: torch/onnx/_internal/registration.py [hunk 1]
```diff
@@ -256,51 +280,66 @@ def all_functions(self) -> Set[str]:
         return set(self._registry)
 
 
-def discover_and_register_all_symbolic_opsets() -> None:
```
Comment (justinchuby) [inline torch/onnx/_internal/registration.py:259]: Removed

File: torch/onnx/symbolic_opset8.py [hunk 2]
```diff
@@ -119,6 +136,7 @@ def __interpolate(
 
 # NOTE: We should create a wrapper for this kind of operation, after resolving the shape/type propagation
 #       issue for "cast" operators. Some symbolic functions depend on shape information of input tensor, which
+
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset8.py:139]: To remove

File: torch/onnx/symbolic_opset8.py [hunk 3]
```diff
@@ -159,15 +177,19 @@ def _comparison_operator(g, input, other, op_name):
 
 
 # NOTE: For symbolics {gt, lt, bmm, matmul, prelu, mm, addmm, view, flatten},
+
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset8.py:180]: Remove

File: torch/onnx/symbolic_opset9.py [hunk 4]
```diff
@@ -5836,323 +6148,330 @@ def broadcast_tensors(g, self):
...
+                new_block,
+                operator_export_type,  # type:ignore[arg-type]
+                env,
+                False,
             )
         new_op_outputs = torch._C._jit_pass_fixup_onnx_controlflow_node(
             new_node, opset_version
         )
-        # Run shape type inference for Loop after subblock is converted.
+        # Run shape type inference for If after subblock is converted.
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset9.py:6432]: git diff issues: this line is line 6108 in the old code

File: torch/onnx/symbolic_caffe2.py [hunk 5]
```diff
@@ -194,7 +194,7 @@ def upsample_nearest2d(
 @symbolic_helper.parse_args("v", "is", "is", "is", "is", "i")
 def max_pool2d(g, input, kernel_size, stride, padding, dilation, ceil_mode):
     if input not in symbolic_helper._quantized_ops:
-        return opset9.max_pool2d(
+        return opset9.max_pool2d(  # type: ignore[attr-defined]
```
Comment (BowenBao) [inline torch/onnx/symbolic_caffe2.py:197]: Curious, why is this needed?

Comment (BowenBao) [reply torch/onnx/symbolic_caffe2.py:197]: I see, these are no longer functions in `symbolic_opset9.py`. I guess this fall should fail?

File: torch/onnx/symbolic_opset9.py [hunk 6]
```diff
@@ -335,40 +312,66 @@
...
+def _apply_params(*args, **kwargs):
+    """Returns a decorator that calls the decorated (higher-order) function with the given parameters."""
+
+    def _apply(fn):
+        return fn(*args, **kwargs)
+
+    return _apply
+
+
+def _export(name: str):
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset9.py:327]: @BowenBao Re: `return opset9.max_pool2d(  # type: ignore[attr-defined]`

The functions are no longer exactly defined but is still (only) called in symbolic_caffe2. I created this _export decorator to expose the names. When we have the dispatcher, we can leverage that instead.

File: torch/onnx/symbolic_opset8.py [hunk 7]
```diff
@@ -30,12 +30,16 @@
...
 
+import functools
 import warnings
 
 import torch
 from torch.onnx import _type_utils, errors, symbolic_helper, symbolic_opset9 as opset9
+from torch.onnx._internal import registration
 
-block_listed_operators = [
+_onnx_symbolic = functools.partial(registration.onnx_symbolic, opset=8)
```
Comment (AllenTiTaiWang) [inline torch/onnx/symbolic_opset8.py:40]: It seems that the version number is duplicated defined here. ex: We have file name telling us it's opset8, and the registry telling us again. Will we have a next step to only depend on decorator opset version number?

Comment (justinchuby) [reply torch/onnx/symbolic_opset8.py:40]: Great question. Yes - with this way of registering we will not depend on how the module is named. I am looking at how we can better structure our functions. Until then the functions will stay in these versioned modules. We need to make them accessible for at least a few release versions for compatibility too. Any thoughts / suggestions?

File: torch/onnx/symbolic_caffe2.py (hunk 5 above)
Comment (justinchuby) [reply torch/onnx/symbolic_caffe2.py:197]: A: https://github.com/pytorch/pytorch/pull/84448#discussion_r966486511

File: torch/onnx/symbolic_opset8.py (hunk 7 above)
Comment (AllenTiTaiWang) [reply torch/onnx/symbolic_opset8.py:40]: I guess without making each symbolic_fn as class, it only matters how we are going to categorize them. As now we don't need to categorize them in opset version way, maybe we can discuss a categorization that we can benefit from the most?

Comment (justinchuby) [reply torch/onnx/symbolic_opset8.py:40]: That sounds like a great plan

**Review by titaiwangms — APPROVED** [review]
LGTM!

File: torch/onnx/_internal/registration.py [hunk 8]
```diff
@@ -256,51 +280,66 @@ def all_functions(self) -> Set[str]:
...
-    """Discover all symbolic functions.
-
-    Opset 9 is the base version. It is selected as the base version because
-        1. It is the first opset version supported by PyTorch export.
-        2. opset 9 is more robust than previous opset versions. Opset versions like 7/8 have limitations
-            that certain basic operators cannot be expressed in ONNX. Instead of basing on these limitations,
-            we chose to handle them as special cases separately.
+@_beartype.beartype
+def onnx_symbolic(
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/_internal/registration.py:264]: Is unit test included?

File: torch/onnx/symbolic_opset9.py (hunk 6 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py:327]: Should we add this to decorator for all those kinds to maintain same behavior?

What happens to `__all__`, should it be kept there if `_export` is added?

File: torch/onnx/README.md [hunk 9]
```diff
@@ -5,3 +5,21 @@ Torch->ONNX converter / exporter.
...
+2. Opset 9 is more robust than previous opset versions. Opset versions like 7/8 have limitations
+    that certain basic operators cannot be expressed in ONNX. Instead of basing on these limitations,
+    we chose to handle them as special cases separately.
+
+Backward support for opset versions beyond opset 7 is not in our roadmap.
+
+For opset versions other than 9, by default they will inherit the symbolic functions defined in
+symbolic_opset9.py.
+
+To extend support for updated operators in different opset versions on top of opset 9,
```
Comment (BowenBao) [inline torch/onnx/README.md:23]: Let's add a few words about how to add new symbolic functions using the new decorator.

File: torch/onnx/symbolic_opset9.py (hunk 6 above)
Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py:327]: I added `_export` to only those that are called internally. I am hoping to reduce the number of exported functions. The only risk is internal CI breakage that we don't know about.

if _export is added, the entry in `__all__` will be kept.

File: torch/onnx/_internal/registration.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/_internal/registration.py:267]: Added in https://github.com/pytorch/pytorch/pull/84709/files#diff-ad2ff738b3486e05884f6b7cee550f1580bcb5410d863b51ab081622bc8928ba

File: torch/onnx/README.md (hunk 9 above)
Comment (justinchuby) [reply torch/onnx/README.md:23]: Done. Added in https://github.com/pytorch/pytorch/pull/85290/files

**Comment by justinchuby** [comment]
@BowenBao PTAL

**Comment by justinchuby** [comment]
@albanD or @seemethere The CI asked that I contact you. This PR applies a new decorator to functions in torch.onnx so it is big.

**Comment by albanD** [comment]
Hey!
This seems to be doing a lot of things as well as adding these decorators. Would it be possible to split this into two independent PRs?

**Comment by justinchuby** [comment]
@albanD thanks for the suggestion. Any changes in the symbolic_opset{n} files are needed to accommodate the new registration decorator. The rest of the changes are simply (1) remove unused code in `registration.py` and (2) moving text from removed functions to README.md. So there is not much to break out from this PR

**Review by BowenBao — APPROVED** [review]
LGTM, a follow up comment on `_export`. I'd suggest adding _export to ensure `__all__` is unchanged. It is safer to avoid any potential backward breakage, until the inter calling of symbolic function is solved in later work.

**Comment by justinchuby** [comment]
> adding _export to ensure `__all__` is unchanged

Done

File: torch/onnx/symbolic_opset9.py (hunk 6 above)
Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py:327]: Done
